server/src/main/java/org/opensearch/search/query/QueryPhase.java

```diff
@@ -446,8 +447,48 @@ protected boolean searchWithCollector(
             boolean hasTimeout
         ) throws IOException {
             // create the top docs collector last when the other collectors are known
-            final TopDocsCollectorContext topDocsFactory = createTopDocsCollectorContext(searchContext, hasFilterCollector);
-            return searchWithCollector(searchContext, searcher, query, collectors, topDocsFactory, hasFilterCollector, hasTimeout);
+            final Optional<QueryCollectorContext> queryCollectorContextOpt = QueryCollectorContextSpecRegistry.getQueryCollectorContextSpec(
```

Comment:
We can move this logic to a helper method to `getQueryCollectorContext` and then just call `searchWithCollector` once
```java
QueryCollectorContext collectorContext = getQueryCollectorContext(searchContext, hasFilterCollector);

    return searchWithCollector(
        searchContext,
        searcher,
        query,
        collectors,
        collectorContext,
        hasFilterCollector,
        hasTimeout
    );
```